src/thread.cr

```diff
@@ -6,92 +6,125 @@ class Thread
   def initialize
-    @current_fiber = uninitialized Fiber
     @func = ->{}
-    @@threads << self
     @th = LibC.pthread_self
+    @main_fiber = Fiber.new
+
+    @@threads.push(self)
   end
 
   def finalize
-    GC.pthread_detach(@th.not_nil!) unless @detached
+    GC.pthread_detach(@th) unless @detached
   end
 
-  def join
-    GC.pthread_join(@th.not_nil!)
-    @detached = true
+  {% if flag?(:android) || flag?(:openbsd) %}
+    # no thread local storage (TLS) for OpenBSD or Android,
+    # we use pthread's specific storage (TSS) instead:
+    @@current_key : LibC::PthreadKeyT
 
-    if exception = @exception
-      raise exception
+    @@current_key = begin
+      ret = LibC.pthread_key_create(out current_key, nil)
+      raise Errno.new("pthread_key_create") unless ret == 0
+      current_key
     end
-  end
-
-  # All threads, so the GC can see them (GC doesn't scan thread locals)
-  # and we can find the current thread on platforms that don't support
-  # thread local storage (eg: OpenBSD)
-  @@threads = Set(Thread).new
-
-  {% if flag?(:openbsd) %}
-    @@main = new
 
+    # Returns the Thread object associated to the running system thread.
     def self.current : Thread
-      if LibC.pthread_main_np == 1
-        return @@main
-      end
-
-      current_thread_id = LibC.pthread_self
-
-      current_thread = @@threads.find do |thread|
-        LibC.pthread_equal(thread.id, current_thread_id) != 0
+      if ptr = LibC.pthread_getspecific(@@current_key)
+        ptr.as(Thread)
+      else
+        raise "BUG: Thread.current returned NULL"
       end
-
-      raise "Error: failed to find current thread" unless current_thread
-      current_thread
     end
 
-    protected def id
-      @th.not_nil!
+    # Associates the Thread object to the running system thread.
+    protected def self.current=(thread : Thread) : Thread
+      ret = LibC.pthread_setspecific(@@current_key, thread.as(Void*))
+      raise Errno.new("pthread_setspecific") unless ret == 0
+      thread
     end
   {% else %}
     @[ThreadLocal]
-    @@current = new
+    @@current : Thread?
 
-    def self.current
-      @@current
+    # Returns the Thread object associated to the running system thread.
+    def self.current : Thread
+      @@current || raise "BUG: Thread.current returned NULL"
+    end
+
+    # Associates the Thread object to the running system thread.
+    protected def self.current=(@@current : Thread) : Thread
     end
   {% end %}
 
+  # Create the thread object for the current thread (aka the main thread of the
+  # process).
+  #
+  # TODO: consider moving to `kernel.cr` or `crystal/main.cr`
+  self.current = new
+
+  def join
+    GC.pthread_join(@th)
+    @detached = true
+
+    if exception = @exception
+      raise exception
+    end
+  end
+
+  # Returns the Fiber representing the thread's main stack.
+  def main_fiber
+    @main_fiber.not_nil!
+  end
+
+  # :nodoc:
+  def scheduler
+    @scheduler ||= Crystal::Scheduler.new(main_fiber)
```

Comment:
So, besides the refactor this PR will allow each thread to have a the current ST scheduler, but the GC support is not yet guaranteed. Ok, we can go in that direction as intermediate stage.